software/contrib/envelope_generator.py

```diff
@@ -0,0 +1,222 @@
+        self.direction = 1
+    
+    def receive_trigger_fall(self):
+        self.direction = 0
+        
+    def change_sustain_mode(self):
+        self.sustain_mode = 1 - self.sustain_mode
+        
+    def change_looping_mode(self):
+        self.looping_mode = 1 - self.looping_mode
+        
+    def copy_digital_input(self):
+        self.din_copy_out.value(din.value())
+
+    def difference(self, a, b):
+        return abs(a - b)
+    
+    def update_increment_factor(self):
+        increment_factor_rising = k1.range(self.max_increment_factor, 512) / 2
+        self.increment_factor = [(increment_factor_rising + 1), (k2.range(self.max_increment_factor, 256) + 1 + (ain.percent(256) * self.max_increment_factor))]
+        
+    def log(self, number):
+        return log(max(number, 1))
+
+    def update_envelope_value(self):
+        #Envelope rising
+        if self.direction == 1:
+            increment = self.difference(self.envelope_value, self.max_output_voltage) / self.increment_factor[0]
+            self.envelope_value += increment
+            if self.difference(self.envelope_value, self.max_output_voltage) <= self.voltage_threshold:
+                self.envelope_value = self.max_output_voltage
+                if self.sustain_mode == 1 and self.looping_mode == 0:
+                    self.direction = 3
+                else:
+                    self.direction = 0
+            else:
+                sleep_ms(self.increment_delay)
+            
+        #Envelope falling
+        elif self.direction == 0:
+            increment = self.difference(0, self.envelope_value) / self.increment_factor[1]
+            self.envelope_value -= increment
+            if self.difference(0, self.envelope_value) <= self.voltage_threshold:
+                self.envelope_value = 0
+                if self.looping_mode == 0:
+                    self.direction = 3
+                else:
+                    self.direction = 1
+            else:
+                sleep_ms(self.increment_delay)
+            
+        #Update CV output to envelope value
+        self.update_output_voltage()
+            
+        
+    def update_output_voltage(self):
+        self.envelope_out.voltage(self.envelope_value)
+        self.envelope_inverted_out.voltage(self.max_output_voltage - self.envelope_value)
+        
+    def update_display(self):
+        if ticks_ms() - self.last_refreshed_display >= self.display_refresh_rate:
+            #Save state to file
+            self.save_state()
+            
+            #Draw slope graph axis
+            oled.hline(self.envelope_display_bounds[0], self.envelope_display_bounds[3], self.envelope_display_bounds[2], 1)
+            oled.vline(self.envelope_display_bounds[0], self.envelope_display_bounds[1], self.envelope_display_bounds[3], 1)
+            oled.vline((self.envelope_display_bounds[2] - 1), self.envelope_display_bounds[1], self.envelope_display_bounds[3], 1)
+            
+            try:
```

Comment:
[optional] Pro tip: you want to keep your try/except as tightly narrowed to the scope of where the exception is thrown. When you have this many lines inside of a try/except, you might have a line throw a ZeroDivisionError that you weren't expecting and that would hide the bug from you and make troubleshooting tricky. Additionally, it's not obvious to the reader where in this block the ZeroDivisionError might get thrown.

Reply:
fixed! hopefully the code is more clear now